File: Data Structures & Algorithms/foreign-dictionary/submission-3.java
class Solution {
    public String foreignDictionary(String[] words) {
      Map<Character, List<Character>> graph = new HashMap<>();
      Map<Character, Integer> indegree = new HashMap<>();
      for (String word : words) {
        for (char c : word.toCharArray()) {
            indegree.putIfAbsent(c, 0);
            graph.putIfAbsent(c, new ArrayList<>());
        }
      }

      for (int i = 0; i < words.length - 1; i++) {
            String first = words[i];
            String second = words[i+1];
            int minLen = Math.min(first.length(), second.length());
            if (
                first.length() > second.length() &&
                first.substring(0, minLen).equals(second.substring(0, minLen))
            ) {
                return "";
            }
            for (int j = 0; j < minLen; j++) {
                char c1 = first.charAt(j);
                char c2 = second.charAt(j);
                if (c1 != c2) {
                    graph.get(c1).add(c2);
                    indegree.put(c2, indegree.get(c2) + 1);
                    break;
                }
            }
        }
        Queue<Character> que = new LinkedList<>();
        for (char c : indegree.keySet()) {
            if (indegree.get(c) == 0) {
                que.offer(c);
            }
        }

        StringBuilder res = new StringBuilder();
        while (!que.isEmpty()) {
            char ch = que.poll();
            res.append(ch);
            for (char nei : graph.getOrDefault(ch, new ArrayList<>())) {
                indegree.put(nei, indegree.get(nei) - 1);
                if (indegree.get(nei) == 0) {
                    que.offer(nei);
                }
            }
        }

        if (res.length() != indegree.size()) {
            return "";
        }

        return res.toString();
    }
}

File: Data Structures & Algorithms/count-paths/submission-3.py
# class Solution:
#     def uniquePaths(self, m: int, n: int) -> int:
#         dp = [1] * n
#         for i in range(1, m):
#             for j in range(1, n):
#                 dp[j] += dp[j-1]
        
#         return dp[n-1]

class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        return self.combination(m+n-2, min(m-1, n-1))
    
    def combination(self, a, b):
        res = 1
        for i in range(1, b+1):
            res = res * (a - i + 1) // i
        
        return res